dango/types/src/lending/interest_rate.rs

```diff
@@ -0,0 +1,259 @@
+    /// is applied when the utilization is above the optimal utilization rate.
+    DualSlope {
+        /// The base interest rate. This is the interest rate that is applied
+        /// when the utilization is 0%.
+        base_rate: Bounded<Udec128, ZeroInclusiveOneExclusive>,
+        /// The optimal utilization rate. This is the utilization rate after
+        /// which the second slope is applied.
+        optimal_utilization: Bounded<Udec128, ZeroExclusiveOneExclusive>,
+        /// The slope of the first linear function. This is the slope that is
+        /// applied when the utilization is below the optimal utilization rate.
+        first_slope: Bounded<Udec128, ZeroExclusiveOneExclusive>,
+        /// The slope of the second linear function. This is the slope that is
+        /// applied when the utilization is above the optimal utilization rate.
+        second_slope: Bounded<Udec128, ZeroExclusiveOneExclusive>,
+        /// The portion of interest retained as protocol reserves.
+        reserve_factor: Bounded<Udec128, ZeroInclusiveOneInclusive>,
+    },
+}
+
+/// Represents the calculated interest rates at a given utilization
+#[derive(Debug)]
+pub struct InterestRates {
+    pub borrow_rate: Udec128,
+    pub deposit_rate: Udec128,
+    pub spread: Udec128,
+}
+
+impl Display for InterestRates {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "borrow_rate: {}, deposit_rate: {}, spread: {}",
+            self.borrow_rate, self.deposit_rate, self.spread
+        )
+    }
+}
+
+impl InterestRateModel {
+    /// Calculates interest rates for a given utilization rate
+    pub fn calculate_rates(&self, utilization: Udec128) -> anyhow::Result<InterestRates> {
+        match self {
+            Self::DualSlope {
+                base_rate,
+                optimal_utilization,
+                first_slope,
+                second_slope,
+                reserve_factor,
+            } => {
+                if utilization > Udec128::new_percent(100) {
+                    return Err(anyhow!("invalid utilization rate"));
+                }
+
+                // Calculate borrow rate
+                let borrow_rate = if utilization <= **optimal_utilization {
+                    // Below optimal: linear increase
+                    **base_rate + (utilization / **optimal_utilization) * **first_slope
+                } else {
+                    // Above optimal: steeper increase
+                    let excess_utilization = (utilization - **optimal_utilization)
+                        / (Udec128::ONE - **optimal_utilization);
+                    **base_rate + **first_slope + (excess_utilization * **second_slope)
+                };
+
+                // Calculate deposit rate
+                let deposit_rate = utilization * borrow_rate * (Udec128::ONE - **reserve_factor);
+
+                // Calculate spread
+                let spread = borrow_rate - deposit_rate;
```

Comment:
Double check these won't panic. Alternatively use checked math everywhere.

Reply:
Should not be possible since values are bounded. But can still convert to checked math if you prefer? Just makes it much harder to read imo.

Reply:
Let's keep it as is now then

Reply:
Let's keep it as is now then